controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1420,6 +1486,89 @@ private TxnStatus handleDataNotFoundException(Throwable ex) {
+                                createWriterMarkRecord(writer, timestamp, newPosition)
+                                        .thenApply(v -> WriterTimestampResponse.SUCCESS),
+                                e -> Exceptions.unwrap(e) instanceof StoreException.DataExistsException, 
+                                () -> noteWriterMark(writer, timestamp, position));
+                    } else {
+                        // sanity check and update
+                        if (record.getObject().getTimestamp() > timestamp) {
+                            // existing time is already ahead of new time
+                            return CompletableFuture.completedFuture(WriterTimestampResponse.INVALID_TIME);
+                        }
+                        if (!compareWriterPositions(record.getObject().getPosition(), newPosition)) {
+                            // existing position is already ahead of new position
+                            return CompletableFuture.completedFuture(WriterTimestampResponse.INVALID_POSITION);
+                        }
+                        
+                        // its a valid mark, update it
+                        return updateWriterMarkRecord(writer, timestamp, newPosition, record.getVersion())
+                                .thenApply(v -> WriterTimestampResponse.SUCCESS);
+                    }
+                });
+    }
+
+
+    /**
+     * Compares two given positions and returns true if position 2 is ahead of position 1. 
+     * 
+     * Note: This is not absolutely correct implementation because that would be a very costly computation. 
+     * This method simply checks if position 2 has a higher max segment number than position 1. 
+     * If those are equal, then it simply checks the offsets in each segment in position 2 to be ahead of same segment in 
+     * position 1. 
+     * It doesnt deal with complex cases where in a position 1 there could be both predecessor and successor while position 2 
+     * only contains successor. There can be many such complex checks which will require looking up ranges of segments involved
+     * which we will avoid in this method. 
+     * 
+     * @param position1 position 1 
+     * @param position2 position 2
+     * @return return true if position 2 is ahead of/or equal to position 1. false otherwise. 
+     */
+    @VisibleForTesting
+    boolean compareWriterPositions(Map<Long, Long> position1, Map<Long, Long> position2) {
+        ArrayList<Map.Entry<Long, Long>> sortedPosition1 = new ArrayList<>(position1.entrySet());
+        sortedPosition1.sort(Collections.reverseOrder(Comparator.comparingLong(Map.Entry::getKey)));
+        ArrayList<Map.Entry<Long, Long>> sortedPosition2 = new ArrayList<>(position2.entrySet());
+        sortedPosition2.sort(Collections.reverseOrder(Comparator.comparingLong(Map.Entry::getKey)));
+
+        long maxInPos2Only = position2.keySet().stream().filter(position1::containsKey).max(Long::compare).orElse(Long.MIN_VALUE);
```

Comment:
nit: "only" isn't really needed in the name.

Reply:
Both this and the next lines are guaranteed to be O(n). You can make this at most O(n) by looping through the already sorted positions (descending) which will prevent you from doing unnecessary comparisons.

Reply:
Actually why do you sort them anyway? You never seem to use `sortedPosition1` or ``sortedPosition2` anywhere.

Reply:
Actually you are right.. I was doing more checks with sorting.. but i realized those checks are inconclusive. I intended to not have the sorted as it is useless.